# Flipkart Product Pricing & Discount Analysis

## Data Understanding & Data Cleaning

### End-to-End Data Analytics Project

---

### Domain
E-Commerce Analytics

### Dataset
Flipkart Product Catalogue (2015–2016)

### Tools & Technologies
- Python
- Pandas
- NumPy
- Matplotlib
- Seaborn

---

## Notebook Objective

This notebook focuses on understanding the Flipkart product catalogue dataset, assessing its quality, cleaning inconsistencies, engineering useful analytical features, and preparing a reliable dataset for exploratory data analysis and dashboard development.

The cleaned dataset generated from this notebook will be used in subsequent notebooks for business analysis and visualization.

# Business Problem

Flipkart hosts thousands of products across multiple categories with varying prices, discounts, brands, and customer ratings.

Without structured analysis, it is difficult to understand:

- how products are distributed across categories,
- how pricing strategies vary,
- whether discounts are consistent,
- and whether catalogue quality issues affect business decisions.

The objective of this project is to analyse the product catalogue to uncover pricing patterns, discount behaviour, catalogue composition, and data quality issues that can support better pricing strategies and catalogue management.

# Project Objectives

The primary objectives of this project are:

- Understand the structure and quality of the dataset.
- Identify and clean missing, inconsistent, and duplicate records.
- Create analytical features that improve business analysis.
- Prepare a reliable dataset for exploratory data analysis.
- Build a clean dataset suitable for Power BI dashboard development.

# Stakeholders

The insights generated from this analysis can support the following stakeholders:

- Category Managers
- Pricing Analysts
- Business Analysts
- Business Intelligence Teams
- Product Managers
- Senior Management

These stakeholders can use the findings to understand catalogue composition, evaluate pricing strategies, identify data quality issues, and support pricing and merchandising decisions.

# Business Questions

This notebook prepares the dataset required to answer the following business questions:

1. Which product categories dominate the catalogue?

2. How are products priced across different categories?

3. Which categories receive the highest discounts?

4. How are products distributed across different price segments?

5. Which brands have the highest catalogue presence?

6. What data quality issues exist within the catalogue?

7. Is the available rating data sufficient for reliable analysis?

# Dataset Overview

**Dataset Name:** Flipkart Product Catalogue

**Source:** Publicly available Flipkart dataset (2015–2016)

### Dataset Size

- Total Records : **20,000**
- Original Columns : **15**

### Columns Selected for Analysis

| Column | Purpose |
|---------|---------|
| brand | Brand analysis |
| product_name | Product identification |
| product_category_tree | Category extraction |
| retail_price | Original product price |
| discounted_price | Selling price |
| product_rating | Customer rating analysis |

### Excluded Columns

The following columns were excluded because they do not contribute directly to solving the selected business problem:

- uniq_id
- crawl_timestamp
- product_url
- pid
- image
- is_FK_Advantage_product
- description
- overall_rating
- product_specifications

# Assumptions & Project Limitations

## Assumptions

- Each row represents a unique product listing.
- Listings with different prices are treated as different product listings even if the product names are identical.
- Retail Price represents the original listed price.
- Discounted Price represents the selling price.
- Missing brand values indicate unavailable information.

---

## Limitations

- The dataset contains catalogue information only.
- No customer transactions are available.
- No sales or revenue data is available.
- No inventory information is available.
- Approximately 90% of products do not contain customer ratings.
- Missing brand information limits brand-level analysis.

# Project Workflow

The project follows the CRISP-DM inspired analytics workflow.

1. Business Understanding

2. Data Understanding

3. Data Cleaning

4. Feature Engineering

5. Export Clean Dataset

6. Exploratory Data Analysis

7. Business Insights

8. Power BI Dashboard

9. Business Recommendations

10. Final Documentation

In [5]:
# ==========================================================
# Import Required Libraries
# ==========================================================

import warnings
warnings.filterwarnings('ignore')

import pandas as pd
import numpy as np

import matplotlib.pyplot as plt
import seaborn as sns

sns.set_theme(style="whitegrid")

plt.rcParams["figure.figsize"] = (8,5)
plt.rcParams["figure.dpi"] = 120

In [6]:
# ==========================================================
# Load Dataset
# ==========================================================

df = pd.read_csv("../data/flipkart_dataset.csv")

print("="*60)
print("Dataset Loaded Successfully")
print("="*60)

print(f"Rows    : {df.shape[0]}")
print(f"Columns : {df.shape[1]}")

df.head()

Dataset Loaded Successfully
Rows    : 20000
Columns : 15


,uniq_id,crawl_timestamp,product_url,product_name,product_category_tree,pid,retail_price,discounted_price,image,is_FK_Advantage_product,description,product_rating,overall_rating,brand,product_specifications
0,c2d766ca982eca8304150849735ffef9,2016-03-25 22:59:23 +0000,http://www.flipkart.com/alisha-solid-women-s-c...,Alisha Solid Women's Cycling Shorts,"[""Clothing >> Women's Clothing >> Lingerie, Sl...",SRTEH2FF9KEDEFGF,999.0,379.0,"[""http://img5a.flixcart.com/image/short/u/4/a/...",False,Key Features of Alisha Solid Women's Cycling S...,No rating available,No rating available,Alisha,"{""product_specification""=>[{""key""=>""Number of ..."
1,7f7036a6d550aaa89d34c77bd39a5e48,2016-03-25 22:59:23 +0000,http://www.flipkart.com/fabhomedecor-fabric-do...,FabHomeDecor Fabric Double Sofa Bed,"[""Furniture >> Living Room Furniture >> Sofa B...",SBEEH3QGU7MFYJFY,32157.0,22646.0,"[""http://img6a.flixcart.com/image/sofa-bed/j/f...",False,FabHomeDecor Fabric Double Sofa Bed (Finish Co...,No rating available,No rating available,FabHomeDecor,"{""product_specification""=>[{""key""=>""Installati..."
2,f449ec65dcbc041b6ae5e6a32717d01b,2016-03-25 22:59:23 +0000,http://www.flipkart.com/aw-bellies/p/itmeh4grg...,AW Bellies,"[""Footwear >> Women's Footwear >> Ballerinas >...",SHOEH4GRSUBJGZXE,999.0,499.0,"[""http://img5a.flixcart.com/image/shoe/7/z/z/r...",False,Key Features of AW Bellies Sandals Wedges Heel...,No rating available,No rating available,AW,"{""product_specification""=>[{""key""=>""Ideal For""..."
3,0973b37acd0c664e3de26e97e5571454,2016-03-25 22:59:23 +0000,http://www.flipkart.com/alisha-solid-women-s-c...,Alisha Solid Women's Cycling Shorts,"[""Clothing >> Women's Clothing >> Lingerie, Sl...",SRTEH2F6HUZMQ6SJ,699.0,267.0,"[""http://img5a.flixcart.com/image/short/6/2/h/...",False,Key Features of Alisha Solid Women's Cycling S...,No rating available,No rating available,Alisha,"{""product_specification""=>[{""key""=>""Number of ..."
4,bc940ea42ee6bef5ac7cea3fb5cfbee7,2016-03-25 22:59:23 +0000,http://www.flipkart.com/sicons-all-purpose-arn...,Sicons All Purpose Arnica Dog Shampoo,"[""Pet Supplies >> Grooming >> Skin & Coat Care...",PSOEH3ZYDMSYARJ5,220.0,210.0,"[""http://img5a.flixcart.com/image/pet-shampoo/...",False,Specifications of Sicons All Purpose Arnica Do...,No rating available,No rating available,Sicons,"{""product_specification""=>[{""key""=>""Pet Type"",..."


# Data Understanding

Before cleaning the dataset, it is important to understand its overall structure, data types, descriptive statistics, missing values, and duplicate records.

This initial assessment helps identify potential data quality issues that may affect the reliability of the analysis.

In [7]:
# ==========================================================
# Select Required Columns
# ==========================================================

df = df[
    [
        "brand",
        "product_name",
        "product_category_tree",
        "product_rating",
        "retail_price",
        "discounted_price",
    ]
]

print(f"Dataset Shape : {df.shape}")

display(df.head())

print("\nDataset Information\n")
display(df.info())

print("\nDescriptive Statistics\n")
display(df.describe())

Dataset Shape : (20000, 6)


,brand,product_name,product_category_tree,product_rating,retail_price,discounted_price
0,Alisha,Alisha Solid Women's Cycling Shorts,"[""Clothing >> Women's Clothing >> Lingerie, Sl...",No rating available,999.0,379.0
1,FabHomeDecor,FabHomeDecor Fabric Double Sofa Bed,"[""Furniture >> Living Room Furniture >> Sofa B...",No rating available,32157.0,22646.0
2,AW,AW Bellies,"[""Footwear >> Women's Footwear >> Ballerinas >...",No rating available,999.0,499.0
3,Alisha,Alisha Solid Women's Cycling Shorts,"[""Clothing >> Women's Clothing >> Lingerie, Sl...",No rating available,699.0,267.0
4,Sicons,Sicons All Purpose Arnica Dog Shampoo,"[""Pet Supplies >> Grooming >> Skin & Coat Care...",No rating available,220.0,210.0



Dataset Information

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20000 entries, 0 to 19999
Data columns (total 6 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   brand                  14136 non-null  object 
 1   product_name           20000 non-null  object 
 2   product_category_tree  20000 non-null  object 
 3   product_rating         20000 non-null  object 
 4   retail_price           19922 non-null  float64
 5   discounted_price       19922 non-null  float64
dtypes: float64(2), object(4)
memory usage: 937.6+ KB


None


Descriptive Statistics



,retail_price,discounted_price
count,19922.000000,19922.000000
mean,2979.206104,1973.401767
std,9009.639341,7333.586040
min,35.000000,35.000000
25%,666.000000,350.000000
50%,1040.000000,550.000000
75%,1999.000000,999.000000
max,571230.000000,571230.000000


In [8]:
# ==========================================================
# Missing Value Analysis
# ==========================================================

missing = pd.DataFrame(
    {
        "Missing Values": df.isnull().sum(),
        "Missing Percentage": (
            df.isnull().sum() / len(df) * 100
        ).round(2),
    }
)

display(missing)

,Missing Values,Missing Percentage
brand,5864,29.32
product_name,0,0.00
product_category_tree,0,0.00
product_rating,0,0.00
retail_price,78,0.39
discounted_price,78,0.39


## Observations

The initial assessment of missing values reveals several important data quality issues.

- **Brand** contains **5,864 missing values (~29%)**, which is too large to remove without losing a significant portion of the dataset. These values will be replaced with **"Unknown"** to preserve the records.

- **Retail Price** and **Discounted Price** each contain **78 missing values (<1%)**. Since pricing is essential for this analysis, these rows will be removed.

- **Product Rating** currently shows no missing values because unavailable ratings are stored as the text **"No rating available"** rather than actual null values. This column will be converted to numeric later, after which these entries will become **NaN**.

In [9]:
# ==========================================================
# Duplicate Analysis
# ==========================================================

exact_duplicates = df.duplicated().sum()

print(f"Exact Duplicate Rows : {exact_duplicates}")

soft_duplicates = df.duplicated(
    subset=[
        "brand",
        "product_name",
        "retail_price",
        "discounted_price",
    ],
    keep=False,
)

print(f"Potential Duplicate Listings : {soft_duplicates.sum()}")

display(
    df[soft_duplicates]
    .sort_values("product_name")
    .head(10)
)

Exact Duplicate Rows : 3737
Potential Duplicate Listings : 5719


,brand,product_name,product_category_tree,product_rating,retail_price,discounted_price
3849,3a AUTOCARE,3a AUTOCARE Car Mat Hyundai Grand i10,"[""Automotive >> Accessories & Spare parts >> C...",No rating available,2250.0,1490.0
5260,3a AUTOCARE,3a AUTOCARE Car Mat Hyundai Grand i10,"[""Automotive >> Accessories & Spare parts >> C...",No rating available,2250.0,1490.0
4214,3a AUTOCARE,3a AUTOCARE Car Mat Maruti Swift Dzire,"[""Automotive >> Accessories & Spare parts >> C...",No rating available,2250.0,1490.0
3904,3a AUTOCARE,3a AUTOCARE Car Mat Maruti Swift Dzire,"[""Automotive >> Accessories & Spare parts >> C...",No rating available,2250.0,1490.0
846,NaN,612 League Girl's Gathered Dress,"[""Clothing >> Kids' Clothing >> Girls Wear >> ...",No rating available,1095.0,766.0
847,NaN,612 League Girl's Gathered Dress,"[""Clothing >> Kids' Clothing >> Girls Wear >> ...",No rating available,1095.0,766.0
16807,925 Silver,925 Silver Alloy Silver Bracelet,"[""Jewellery >> Bangles, Bracelets & Armlets >>...",No rating available,1599.0,1498.0
16806,925 Silver,925 Silver Alloy Silver Bracelet,"[""Jewellery >> Bangles, Bracelets & Armlets >>...",No rating available,1599.0,1498.0
16805,925 Silver,925 Silver Alloy Silver Bracelet,"[""Jewellery >> Bangles, Bracelets & Armlets >>...",No rating available,1599.0,1498.0
16804,925 Silver,925 Silver Alloy Silver Bracelet,"[""Jewellery >> Bangles, Bracelets & Armlets >>...",No rating available,1599.0,1498.0


## Duplicate Handling Decision

The dataset contains both exact duplicates and repeated product listings.

After inspecting the repeated records, the following strategy was adopted:

- Exact duplicate rows are removed because they contain identical information.
- Products having different retail or discounted prices are retained since they likely represent different product variants or listing updates.
- Duplicate removal is performed using **brand**, **product name**, **retail price**, and **discounted price**, ensuring that meaningful pricing variations are preserved while redundant records are removed.

In [10]:
# ==========================================================
# Data Cleaning
# ==========================================================

# Convert rating column to numeric
df["product_rating"] = pd.to_numeric(
    df["product_rating"],
    errors="coerce",
)

# Convert prices to numeric
df["retail_price"] = pd.to_numeric(
    df["retail_price"],
    errors="coerce",
)

df["discounted_price"] = pd.to_numeric(
    df["discounted_price"],
    errors="coerce",
)

# Remove rows with missing prices
df = df[
    df["retail_price"].notna()
    & df["discounted_price"].notna()
]

# Fill missing brands
df["brand"] = df["brand"].fillna("Unknown")

# Remove duplicate listings
df = df.drop_duplicates(
    subset=[
        "brand",
        "product_name",
        "retail_price",
        "discounted_price",
    ]
)

print("Cleaning Completed Successfully.")

print(f"Final Dataset Shape : {df.shape}")

display(df.isnull().sum())

Cleaning Completed Successfully.
Final Dataset Shape : (15827, 6)


brand                        0
product_name                 0
product_category_tree        0
product_rating           14207
retail_price                 0
discounted_price             0
dtype: int64

In [11]:
# ==========================================================
# Feature Engineering
# ==========================================================

# Discount Percentage
df["discount_percentage"] = (
    (
        df["retail_price"]
        - df["discounted_price"]
    )
    / df["retail_price"]
) * 100

# Main Category
df["main_category"] = (
    df["product_category_tree"]
    .str.replace(r'[\[\]"]', "", regex=True)
    .str.split(">>")
    .str[0]
    .str.strip()
)

# Sub Category
df["sub_category"] = (
    df["product_category_tree"]
    .str.replace(r'[\[\]"]', "", regex=True)
    .str.split(">>")
    .str[1]
    .str.strip()
)

# Price Bucket
df["price_bucket"] = pd.cut(
    df["discounted_price"],
    bins=[0, 500, 1000, 2000, 5000, 600000],
    labels=[
        "<₹500",
        "₹500–1K",
        "₹1K–2K",
        "₹2K–5K",
        "₹5K+",
    ],
)

display(df.head())

,brand,product_name,product_category_tree,product_rating,retail_price,discounted_price,discount_percentage,main_category,sub_category,price_bucket
0,Alisha,Alisha Solid Women's Cycling Shorts,"[""Clothing >> Women's Clothing >> Lingerie, Sl...",NaN,999.0,379.0,62.062062,Clothing,Women's Clothing,<₹500
1,FabHomeDecor,FabHomeDecor Fabric Double Sofa Bed,"[""Furniture >> Living Room Furniture >> Sofa B...",NaN,32157.0,22646.0,29.576764,Furniture,Living Room Furniture,₹5K+
2,AW,AW Bellies,"[""Footwear >> Women's Footwear >> Ballerinas >...",NaN,999.0,499.0,50.050050,Footwear,Women's Footwear,<₹500
3,Alisha,Alisha Solid Women's Cycling Shorts,"[""Clothing >> Women's Clothing >> Lingerie, Sl...",NaN,699.0,267.0,61.802575,Clothing,Women's Clothing,<₹500
4,Sicons,Sicons All Purpose Arnica Dog Shampoo,"[""Pet Supplies >> Grooming >> Skin & Coat Care...",NaN,220.0,210.0,4.545455,Pet Supplies,Grooming,<₹500


In [13]:
# ==========================================================
# Export Clean Dataset
# ==========================================================

df.to_csv(
    "../data/cleaned_flipkart_dataset.csv",
    index=False,
)

print("Cleaned dataset exported successfully.")

print(f"Final Rows    : {df.shape[0]}")
print(f"Final Columns : {df.shape[1]}")

Cleaned dataset exported successfully.
Final Rows    : 15827
Final Columns : 10


# Cleaning Summary

The dataset has been successfully prepared for exploratory data analysis.

## Summary of Cleaning Performed

- Selected only the columns required for the business problem.
- Converted the product rating column to numeric format.
- Removed rows with missing retail and discounted prices.
- Replaced missing brand values with **"Unknown"**.
- Removed duplicate product listings while preserving meaningful pricing variations.
- Created new analytical features:
  - Discount Percentage
  - Main Category
  - Sub Category
  - Price Bucket
- Exported the cleaned dataset for downstream analysis.

## Final Dataset

- **Rows:** 15,827
- **Columns:** 10

The cleaned dataset is now ready for exploratory data analysis in the next notebook.